src/common/common-thread-private.h

```diff
@@ -123,6 +123,61 @@ int COMMON_PREFIX (thread_create) (bson_thread_t *thread,
 bool COMMON_PREFIX (mutex_is_locked) (bson_mutex_t *mutex);
 #endif
 
+/**
+ * @brief A shared mutex (a read-write lock)
+ *
+ * A shared mutex can be locked in 'shared' mode or 'exclusive' mode. Only one
+ * thread may hold exclusive mode at a time. Any number of threads may hold
+ * the lock in shared mode simultaneously. No thread can hold in exclusive mode
+ * while another thread holds in shared mode, and vice-versa.
+ */
+typedef struct bson_shared_mutex_t {
+   BSON_IF_WINDOWS (SRWLOCK native;)
+   BSON_IF_POSIX (pthread_rwlock_t native;)
```

Comment:
to consider: Note that pthreads also has a pthread_mutex_t, which makes me wonder if bson_shared_mutex_t is the right name, especially as the platforms also make that distinction. bson_shared_rwlock? "Archibald"?